README.rst
django-tabular-permissions
##########################
Display Django basic permissions (add, change & delete) in a tabular format that is user friendly, translatable and easy customized.
If you have more customised permissions, the default `FilteredSelectMultiple` widget will also appear to handle those permissions.

*Scroll down for some screen shots.*

Features:
---------
* Permissions are displayed in the active language with app and model ``verbose_name``.
* Any extra custom permissions are displayed in the default admin widget for permissions.
* Customize which apps, models to show in the permissions table. You can also set a exclude function for high-end customization.
* RTL ready, Bootstrap ready.
* Easy customize-able look.
* Tested on Django 1.8 & Django 1.9.
* Tested on Python 2.7, 3.4 & 3.5

<<<<<<< HEAD
Compatibility:
--------------
Django tabular permissions current version 1.1 supports only django 1.11
For earlier versions of django use django-tabular-permissions 1.0.9.




=======
>>>>>>> parent of 005082d... Version 1.10. Support for django 1.11
.. image:: https://travis-ci.org/RamezIssac/django-tabular-permissions.svg?branch=master
    :target: https://travis-ci.org/RamezIssac/django-tabular-permissions


Installation
------------
You can install `django-tabular-permissions` via Pypi::

    pip install django-tabular-permissions

Usage:
------
Simply add "tabular_permissions" to your INSTALLED_APPS setting like this::

    INSTALLED_APPS = [
        ...
        'tabular_permissions',
    ]

then navigate to User and/or Group change form to see django-tabular-permissions in action.

Settings:
---------

* ``TABULAR_PERMISSIONS_EXCLUDE``

Control which apps, models to show in the permissions table.
By default tabular_permissions exclude contrib apps ``sessions`` , ``contenttypes`` and ``admin`` apps from 
showing their models in the permissions table.
You can add (or override) those apps and/or specify models to exclude, like this::

    TABULAR_PERMISSIONS_EXCLUDE = {
        'app': [],
        'model': [],
        'function': 'tabular_permissions.helpers.TabularPermissionDefaultExcludeFunction', 
                     # A dotted path to a class that implement ``__call__`` that takes model as an argument.
        'override': False # Set to True to override default behavior.
    }    


* ``TABULAR_PERMISSIONS_AUTO_IMPLEMENT``

By default, just by including `tabular_permissions` in your installed_apps, the ``django.contrib.admin.UserAdmin`` (and ``GroupAdmin``) are "patched" to include the tabular_permissions widget.
If you have a custom UserAdmin, then set this option to False and make sure you either:

1. Inherit from `TabularPermissionsUserAdmin` for User admin and from``TabularPermissionsGroupAdmin`` for group admin
2. Inherit from ``UserTabularPermissionsAdminBase`` and ``GroupTabularPermissionsAdminBase`` before admin.ModelAdmin or UserAdmin/GroupAdmin,
3. Set the user_permissions widget to ``tabular_permissions.widgets.TabularPermissionsWidget`` and remember to send a 3rd argument 'permissions' when in Group admin. See ``tabular_permissions.admin`` for information.


* ``TABULAR_PERMISSIONS_TEMPLATE``

Default to 'tabular_permissions/admin/tabular_permissions.html`.
You can either extend or override this template for maximum control.

* ``TABULAR_PERMISSIONS_USE_FOR_CONCRETE``

Default `True`. Till now (Feb 1 2016 - django 1.9), there is an inconsistency with proxy models permissions (ticket `11154 <https://code.djangoproject.com/ticket/11154>`_)
So in case you have proxy models and you create their permissions by hand (via this `gist <https://gist.github.com/magopian/7543724>`_ maybe)
Turn off this option in order to correctly assign your newly created permissions via django-tabular-permissions widget.


JavaScript:
-----------
Located at 'static/tabular_permissions/tabular_permissions.js', it have 3 responsibilities:

1. Upon form submit, the checked permissions in the table are dynamically appended to the form default permission input 
   so the backend can carry on its functionality normally and correctly. 
2. Add handlers for column and row select-all checkboxes.
3. Add a class 'related-widget-wrapper-user-permissions' to the div.related-widget-wrapper
   that contains the table, it serves when you need to manipulate the table container.


Screenshots:
------------
Basic Demo

.. image:: http://i.imgsafe.org/c851707.jpeg
    :target: http://i.imgsafe.org/c851707.jpeg
    :alt: Basic demo

With Custom permission behaviour

.. image:: http://i.imgsafe.org/c506554.jpeg
    :target: http://i.imgsafe.org/c506554.jpeg
    :alt: With Custom permission behaviour

RTL and localized

.. image:: http://i.imgsafe.org/4892b01.jpeg
    :target: http://i.imgsafe.org/4892b01.jpeg
    :alt: RTL and localized

-------

Enjoy and feel free to report any bugs or make pull requests.

CHANGELOG.rst
<<<<<<< HEAD
* 1.1: Support django 1.11
* 1.0.9: Minor improve, use django's `import_string` instead of own hacky code to load the exclude function
=======
* 1.0.9: Minor improve, use django's import_string instead of own hacky code to load the exclude function
>>>>>>> parent of 005082d... Version 1.10. Support for django 1.11
* 1.0.8: Fix django version check, add option for dealing with proxy models permissions
* 1.0.7: fixed Issue#2 , 2 models with the same name in different apps + logic enhancement.
* 1.0.5: made TABULAR_PERMISSIONS_EXCLUDE model and app list case insensitive;
  Handle case where excluded model that does not implement the default permissions;
  Fixes around setup.py
* 1.0.3: Fix RTL, move to 'All' instead of 'Select all' , natively translated by django
* 1.0.2: Added 'Select All' for rows.
* 1.0.1: Added tests, travis CI
* 1.0.0: initial concept proof

tabular_permissions/widgets.py
from __future__ import unicode_literals
from django import __version__ as django_version
from django.apps import apps
from django.contrib.auth.models import Permission
from django.contrib.admin.templatetags.admin_static import static
from django.contrib.admin.widgets import FilteredSelectMultiple
from django.contrib.contenttypes.models import ContentType
from django.template.loader import get_template
from django.utils.encoding import force_text
from django.utils.html import escapejs
from django.utils.safestring import mark_safe
from .app_settings import TABULAR_PERMISSIONS_EXCLUDE_FUNCTION, TABULAR_PERMISSIONS_EXCLUDE_APPS, \
    TABULAR_PERMISSIONS_EXCLUDE_MODELS, TABULAR_PERMISSIONS_TEMPLATE, TABULAR_PERMISSIONS_USE_FOR_CONCRETE
from .helpers import get_perm_name


class TabularPermissionsWidget(FilteredSelectMultiple):
    class Media:
        js = ('tabular_permissions/tabular_permissions.js',)

    def __init__(self, verbose_name, is_stacked, input_name='user_permissions', attrs=None, choices=()):
        super(TabularPermissionsWidget, self).__init__(verbose_name, is_stacked, attrs, choices)
        self.managed_perms = []
        self.input_name = input_name  # in case of UserAdmin, it's 'user_permissions', GroupAdmin it's 'permissions'
        self.hide_original = True

    def render(self, name, value, attrs=None, choices=()):

        apps_available = []  # main container to send to template
        user_permissions = Permission.objects.filter(id__in=value or []).values_list('id', flat=True)
        all_perms = Permission.objects.all().values('id', 'codename', 'content_type_id').order_by('codename')
        excluded_perms = set([])
        codename_id_map = {}
        for p in all_perms:
            codename_id_map['%s_%s' % (p['codename'], p['content_type_id'])] = p['id']

        reminder_perms = codename_id_map.copy()
        # used to detect if the tabular permissions covers all permissions, if so, we don't need to make it visible.

        for app in apps.get_app_configs():
            app_dict = {'verbose_name': force_text(app.verbose_name),
                        'models': []}

            for model_name in app.models:
                model = app.models[model_name]
                ct_id = ContentType.objects.get_for_model(model, for_concrete_model=TABULAR_PERMISSIONS_USE_FOR_CONCRETE).pk
                add_perm_name = get_perm_name(model_name, 'add')
                change_perm_name = get_perm_name(model_name, 'change')
                delete_perm_name = get_perm_name(model_name, 'delete')
                view_perm_name = get_perm_name(model_name, 'view')

                add_perm_id = codename_id_map.get('%s_%s' % (add_perm_name, ct_id), False)
                change_perm_id = codename_id_map.get('%s_%s' % (change_perm_name, ct_id), False)
                delete_perm_id = codename_id_map.get('%s_%s' % (delete_perm_name, ct_id), False)
                view_perm_id = codename_id_map.get('%s_%s' % (view_perm_name, ct_id), False)

                if add_perm_id and change_perm_id and delete_perm_id and view_perm_id and not {
                   add_perm_id, change_perm_id, delete_perm_id, view_perm_id} & excluded_perms:
                    excluded_perms.update([add_perm_id, change_perm_id, delete_perm_id, view_perm_id])
                    reminder_perms.pop('%s_%s' % (add_perm_name, ct_id))
                    reminder_perms.pop('%s_%s' % (change_perm_name, ct_id))
                    reminder_perms.pop('%s_%s' % (delete_perm_name, ct_id))
                    reminder_perms.pop('%s_%s' % (view_perm_name, ct_id))

                    if app.label in TABULAR_PERMISSIONS_EXCLUDE_APPS \
                            or model_name in TABULAR_PERMISSIONS_EXCLUDE_MODELS \
                            or TABULAR_PERMISSIONS_EXCLUDE_FUNCTION()(model):
                        continue

                    app_dict['models'].append({
                        'model_name': model_name,
                        'model': model,
                        'verbose_name_plural': force_text(model._meta.verbose_name_plural),
                        'verbose_name': force_text(model._meta.verbose_name),
                        'add_perm_id': add_perm_id,
                        'add_perm_name': add_perm_name,
                        'change_perm_id': change_perm_id,
                        'change_perm_name': change_perm_name,
                        'delete_perm_id': delete_perm_id,
                        'delete_perm_name': delete_perm_name,
                        'view_perm_id': view_perm_id,
                        'view_perm_name': view_perm_name,
                    })

            if app.models:
                apps_available.append(app_dict)

        request_context = {'apps_available': apps_available, 'user_permissions': user_permissions,
                           'codename_id_map': codename_id_map, 'input_name': self.input_name}
        body = get_template(TABULAR_PERMISSIONS_TEMPLATE).render(request_context).encode("utf-8")
        self.managed_perms = excluded_perms
        if reminder_perms:
            self.hide_original = False

        # Get "original" FilteredSelectMultiple, and hide it if necessary.
        # Next block is a "copy" of FilteredSelectMultiple render(), except the if reminder_perms: check.
        # Due to change in how SelectFilter take its arguments and the dropping of static('admin/') in django1.9
        # there a check on django version

        if attrs is None:
            attrs = {}
        attrs['class'] = 'selectfilter'
        if self.is_stacked:
            attrs['class'] += 'stacked'

        output = [super(FilteredSelectMultiple, self).render(name, value, attrs, choices)]
        if reminder_perms:
            output.append('<script type="text/javascript">addEvent(window, "load", function(e) {')

            if '1.8' in django_version:
                output.append('SelectFilter.init("id_%s", "%s", %s, "%s"); });</script>\n'
                              % (name, self.verbose_name.replace('"', '\\"'), int(self.is_stacked), static('admin/')))
            else:  # 1.9
                output.append('SelectFilter.init("id_%s", "%s", %s); });</script>\n'
                              % (name, escapejs(self.verbose_name), int(self.is_stacked)))

        initial = mark_safe(''.join(output))
        response = ' <hr/>'.join([force_text(body), force_text(initial)])
        return mark_safe(response)

    def render_option(self, selected_choices, option_value, option_label):
        if option_value in self.managed_perms:
            # permission is covered in table, skip it.
            return ''
        self.hide_original = False
        return super(TabularPermissionsWidget, self).render_option(selected_choices, option_value, option_label)

    def build_attrs(self, extra_attrs=None, **kwargs):
        if self.hide_original:
            extra_attrs['style'] = " display:none "
        return super(TabularPermissionsWidget, self).build_attrs(extra_attrs, **kwargs)
